Singularity handling for plots improved

povinsahu1909 opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->
Singularity handling for plots was improved

#### References to other Issues or PRs

Fixes #18455 and Fixes #18247 

#### Brief description of what is fixed or changed
For issue #18455 singularity for the expression was calculated before passing the expression to plot function and the given range was broken into a set of range and passed to the function and expressions are plotted in with a new set of range as multiple plots. 

For issue #18247 smart bound are removed to avoid breaking of y-axis.
#### Other comments
This applies to all the expression which is having roots of denominator which is generating improper graphs with extra lines. But for expression with no roots of denominator it works same as it was working currently.

#### Release Notes

<!-- BEGIN RELEASE NOTES -->
* plotting
  * Singularity handing for plot function improved
<!-- END RELEASE NOTES -->

povinsahu1909 committed:
parsing input according to point of discontinuity for improving singularity handling for plots

povinsahu1909 committed:
discontuinity for non-rational function implemented

povinsahu1909 committed:
error for pi was resolved

povinsahu1909 commented on sympy/plotting/plot.py:
```diff
@@ -1062,8 +1062,8 @@ def __init__(self, parent):
-                self.ax[i].spines['left'].set_smart_bounds(True)
-                self.ax[i].spines['bottom'].set_smart_bounds(False)
+                # self.ax[i].spines['left'].set_smart_bounds(True)
+                # self.ax[i].spines['bottom'].set_smart_bounds(False)
```
for generating continuous y-axis which crosses the x-axis.

Commit:
add Line.bisectors

povinsahu1909 committed:
changed to plot discontinuous expression

mohitacecode commented:
Some imports are unused you can remove them

Smit-create commented on sympy/plotting/plot.py:
```diff
@@ -1536,6 +1536,11 @@ def plot(*args, **kwargs):
+    from sympy.solvers import solve
+    from sympy import real_roots
+    from sympy import simplify
+
```
Please move these imports at the top of the file.

Smit-create commented on sympy/plotting/plot.py:
```diff
@@ -1551,8 +1556,31 @@ def plot(*args, **kwargs):
+    for arg in plot_expr:
+        expr = arg[0]
+        limit = arg[1]
+        frac = fraction(expr) #to find the denominator
```
I don't see any use of `fra`. This can be removed

Smit-create commented:
Please at the relevant test cases of the changes you have made.

povinsahu1909 committed:
improved to plot discontinuous functions

povinsahu1909 commented:
> Please at the relevant test cases of the changes you have made.

Yeah, sure I will add the test cases for the changes.

povinsahu1909 commented:
Is this PR require any more changes?

povinsahu1909 commented:
@sylee957 can you please review this PR for any changes.

Smit-create commented:
Can you please add the tests?

asmeurer commented on sympy/plotting/plot.py:
```diff
@@ -1062,8 +1064,8 @@ def __init__(self, parent):
-                self.ax[i].spines['left'].set_smart_bounds(True)
-                self.ax[i].spines['bottom'].set_smart_bounds(False)
+                # self.ax[i].spines['left'].set_smart_bounds(True)
+                # self.ax[i].spines['bottom'].set_smart_bounds(False)
```
Why are these lines commented out? If they are not needed any more they should be removed.

povinsahu1909 replied:
These two lines are deprecated in Matplotlib 3.2 and will be removed in later releases. And this set_smart_bounds method is also causing issue #18247.
I was confused about whether to remove it or not.

asmeurer commented on sympy/plotting/plot.py:
```diff
@@ -32,6 +32,8 @@
 from sympy.utilities.iterables import is_sequence
 from .experimental_lambdify import (vectorized_lambdify, lambdify)
+from sympy import fraction
+from sympy import real_roots
```
You can add these to the main sympy import [above](https://github.com/sympy/sympy/pull/18856/files#diff-e4ca0e0fed60e3bd29d26591f5ba3f88R29)

povinsahu1909 replied:
Yeah sure I will this

asmeurer commented on sympy/plotting/plot.py:
```diff
@@ -1551,8 +1554,30 @@ def plot(*args, **kwargs):
+        limit = arg[1]
+        frac = fraction(expr) #to find the denominator
+        new_deno = frac[1]
+        if (not new_deno.is_Integer) and (not new_deno.is_Float) and (not new_deno.is_irrational):
```
Why does it matter if they are irrational?

povinsahu1909 replied:
Consider the following function `sin(x) / pi`, as in this function denominator is constant and was an irrational number. So for such situation, this was added.

sylee957 committed:
Use gaussian elimination for positive-definite check

povinsahu1909 committed:
added test for function with singularities

povinsahu1909 committed:
added test for functions with singularities

smichr commented:
> Fixes #18455 and #18247

you have to put the keyword (like "fixes") directly before each `#` or else they won't all close

smichr commented on sympy/plotting/plot.py:
```diff
@@ -1555,8 +1554,30 @@ def plot(*args, **kwargs):
+    for arg in plot_expr:
+        expr = arg[0]
+        limit = arg[1]
+        frac = fraction(expr) #to find the denominator
```
Perhaps you could use `sympy.solvers.denoms` to give you all the denominators. `fraction` is naive; I think you should use `.as_numer_denom()[1]` or `denoms`:
```python
>>> from sympy.solvers.solvers import denoms
>>> denoms(1+2/x,x)
{x}
>>> list(map(lambda e: denoms(e, x), [1+2/x,1/(3-x)]))
[{x}, {3 - x}]
```

povinsahu1909 replied:
I used `fraction` to get the denominator as an expression so that roots can be calculated directly by using `real_roots` method. But if we use `denoms` we have to iterate over each term to calculate root. Will it be effective to do.

smichr replied:
`denoms` looks for all denominators. Won't a singularity anywhere cause a problem? Try your current code on the expression `1 + 2/x`.

povinsahu1909 replied:
`1 + 2/x` fails to produce the desired result so I am replacing `fraction` with `denoms` to calculate all denominators. And will make necessary changes to code

smichr replied:
Doesn't the calculus module already have a singularities function?

smichr replied:
hmm...but it will fail for non-rational functions since, for example,
```python
>>> singularities(1/sqrt(x),x)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/runner/.local/share/virtualenvs/python3/lib/python3.8/site-pac
kages/sympy/calculus/singularities.py", line 84, in singularities
    raise NotImplementedError(
NotImplementedError: Algorithms finding singularities for non-rational functions are not yet implemented.
```

povinsahu1909 replied:
> hmm...but it will fail for non-rational functions since, for example,
> 
> ```python
> >>> singularities(1/sqrt(x),x)
> Traceback (most recent call last):
>   File "<stdin>", line 1, in <module>
>   File "/home/runner/.local/share/virtualenvs/python3/lib/python3.8/site-pac
> kages/sympy/calculus/singularities.py", line 84, in singularities
>     raise NotImplementedError(
> NotImplementedError: Algorithms finding singularities for non-rational functions are not yet implemented.
> ```

For finding the singularity we can use `denoms` and then calculate root of the each term of denominator using `real_roots`, but on applying `real_roots` function on a periodic function it is returning a single root.

smichr replied:
yes...`real_roots(sqrt(x - 2) - 1) -> [1]` because `real_roots` is looking at the generator of `Poly(sqrt(x - 2) -1)` which is `sqrt(x - 2)`. So it means there will be a root when `sqrt(x - 2) = 1`. It does not try to solve for the value of `x`.

sylee957 committed:
Resolve merge conflict

povinsahu1909 committed:
singularity calculated term wise for function

smichr commented on sympy/plotting/plot.py:
```diff
@@ -1559,21 +1560,23 @@ def plot(*args, **kwargs):
-        if (not new_deno.is_Integer) and (not new_deno.is_Float) and (not new_deno.is_irrational):
-            root_deno = real_roots(new_deno)   #to find the points of discontinuity
+        root_deno = []
+        for exp in denoms(expr):
```
use `denoms(expr, x)` where x is `expr.free_symbols().pop()` -- assuming a univariate expression -- and then only denominators that contain the symbol of interest will be returned

povinsahu1909 replied:
okay I'll change this

smichr commented on sympy/plotting/plot.py:
```diff
@@ -1559,21 +1560,23 @@ def plot(*args, **kwargs):
-        if (not new_deno.is_Integer) and (not new_deno.is_Float) and (not new_deno.is_irrational):
-            root_deno = real_roots(new_deno)   #to find the points of discontinuity
+        root_deno = []
+        for exp in denoms(expr):
```
It would be nice to use `d` or `den` instead of `exp` which looks like SymPy function `exp()`

povinsahu1909 replied:
I'll replace this

povinsahu1909 committed:
getting denominator with interested symbol

smichr commented:
I am not sure why the passing build is nor reporting back here properly.

smichr commented on sympy/plotting/plot.py:
```diff
@@ -1555,8 +1555,28 @@ def plot(*args, **kwargs):
+        _, x1, x2 = limit
+        root_deno = []
+        for den in denoms(expr, x):
+            root_deno.extend(real_roots(den))
```
`real_roots` is only going to work with polynomial `den`; something like `sqrt(x - 2)` or `sin(x)` will not work. Perhaps `singularities` can be made more robust for these cases.

povinsahu1909 replied:
Instead of `real_roots` can we opt for `solve` to calculate the roots, and considering only real root from the obtained list.
As after changing `real_roots` by `solve` locally I tried plotting `1/(sqrt(x - 2) - 1)` and got the correct output.

povinsahu1909 replied:
`solve` method generates root in terms of `pi` also. But while passing `pi` as a limit in the function it is not generating proper output. Is there is any way to do it?

smichr replied:
evaluate it (e.g. `pi.n()` will give a float that starts as  "3.141592..."; if you want fewer digits you can use `pi.n(3)` to get 3 sig figs or `pi.round(3)` to round to the 1/1000 place (approx).

smichr replied:
`real_roots` is great if the expression is polynomial in `x`; if not then maybe you could fall back to `solve`.

smichr replied:
You would have to watch out for periodic functions in denominators since more than 1 period might appear in the range of interest and `solve` doesn't give you all values in a given range. But the `periodicity` function can supply that information. `solveset`, too, might be the way to go: e.g if plotting x/(sin(2*x) - 1/2) in the range of [3,12]
```python
>>> [i.n(3) for i in solveset(sin(2*x)-S.Half,domain=Interval(3,12))]
[3.40, 4.45, 7.59, 6.54, 9.69, 10.7]
```

smichr replied:
/!\ solutions are not given if that S.Half is replaced with 0.5, however.

povinsahu1909 replied:
> /!\ solutions are not given if that S.Half is replaced with 0.5, however.

Can we use `nsimplify` to convert float in expression into rational form and then passing it in `solveset`.
```
>>> [i.n(3) for i in solveset(nsimplify(sin(2*x)-S.Half),domain=Interval(3,12))]
[3.40, 4.45, 7.59, 6.54, 9.69, 10.7]
>>> [i.n(3) for i in solveset(nsimplify(sin(2*x)-1/2),domain=Interval(3,12))]
[3.40, 4.45, 7.59, 6.54, 9.69, 10.7]
>>> [i.n(3) for i in solveset(nsimplify(sin(2*x)-0.5),domain=Interval(3,12))]
[3.40, 4.45, 7.59, 6.54, 9.69, 10.7]

smitgajjar committed:
parsing(c_parser): added support for binary operators

- Added support for parsing binary operators `+`, `-`, `*`, `/`, `%`, `==`, `!=`, `<`, `<=`, `>`, `>=`, `&&` and `||`
- Added support for parsing assignment(`=`) from one variable to another
- Added support for assignment of integer as well as floating point literal to a variable
- Added support for an rhs expression comprising of any combination of the above operators
- Any form of parenthesised rhs expression will also be parsed successfully
- Added support for all forms of corresponding variable declaration as well as assignment
- Added support for recognizing boolean literals(`true` and `false`) in variable declaration as well as assignment
- Added support for boolean data type
- Corresponding tests are added

smichr commented:
I am collaborating on #19024 right now and you might want to wait until changes from there have been merged (or get the PR and rebase this on top of that). You could be able to use the continuous domain description to create the sub-plots.
```python
for i in continuous_domain(expr, x, Interval(xmin, xmax)):
    plots.append((expr,(x, i.inf, i.sup))
```
...or something like that.

povinsahu1909 committed:
used solveset method

povinsahu1909 commented:
> I am collaborating on #19024 right now and you might want to wait until changes from there have been merged (or get the PR and rebase this on top of that). You could be able to use the continuous domain description to create the sub-plots.
> 
> ```python
> for i in continuous_domain(expr, x, Interval(xmin, xmax)):
>     plots.append((expr,(x, i.inf, i.sup))
> ```
> 
> ...or something like that.

I am working on it

smichr commented:
master branch now has upgraded `continous_domain` which is going to do most of your work by giving you the continuous regions in the plot range with singularities and such removed, e.g. `continuous_domain(sqrt(x - 1/abs(x -3)), x, Interval(-10,10))` will give a Union of the Intervals in which real values for the expression are defined.

smichr commented:
I don't know why results are not being reported back here. #19024 never updated and I just had to open a new PR. I sent a request for support to travis and github.

povinsahu1909 commented:
> I am collaborating on #19024 right now and you might want to wait until changes from there have been merged (or get the PR and rebase this on top of that). You could be able to use the continuous domain description to create the sub-plots.
> 
> ```python
> for i in continuous_domain(expr, x, Interval(xmin, xmax)):
>     plots.append((expr,(x, i.inf, i.sup))
> ```
> 
> ...or something like that.

```
continuous_domain(Integral(log((sin(x)**2 + 1)*sqrt(x**2 + 1)), (x, 0, y)) , y, Interval(1,5))
``` 
running this is producing value error.
```
ValueError: 
sqrt(x**2 + 1)*(sin(x)**2 + 1) > 0 contains imaginary parts which
cannot be made 0 for any value of y satisfying the inequality, leading
to relations like I < 0.
```

smichr commented:
looks like the domain is not being handled somewhere

povinsahu1909 commented:
> looks like the domain is not being handled somewhere
> ...
> update: send x (the variable of integration) rather than the upper limit. With the lower at 0 and upper at y (going from 1 to 5) the x range to check is Interval(0,5)
> 
> ```python
> In [1]: continuous_domain(Integral(log((sin(x)**2 + 1)*sqrt(x**2 + 1)),(x,0,y))
>    ...: ,x,Interval(0,5))
> Out[1]: [0, 5]
> ```

But replacing symbol will not give the correct result, because the obtained function from integral was in terms of y, so if we replace the symbol with other terms it will give result according to symbol passed, 
for example,
```python
>>> continuous_domain(Integral (x/y,x, (x,0,y)) , y, Interval(-1,5))
Union(Interval.Ropen(-1, 0), Interval.Lopen(0, 5))

>>>continuous_domain(Integral(x/y,x, (x,0,y)) , x, Interval(-1,5))
Interval(-1, 5)
```
In the above example changing the symbol, will change the continuous domain.

smichr commented:
> In the above example changing the symbol, will change the continuous domain.

looks like in this example the univariate expression will require two calls to get the domains and the final domain will be their intersection.

smichr commented:
> when function contains imaginary zeros.

Could you give me a univariate example?

povinsahu1909 commented:
```python
>>>continuous_domain(Integral(log(x**3+1),(x,0,y)), y ,Interval(-10,5))
ValueError: 
x**3 + 1 > 0 contains imaginary parts which cannot be made 0 for any
value of y satisfying the inequality, leading to relations like I < 0.

>>> p = Integral(log(x**3+1),(x,0,y)).doit()

>>>p
y*log(y**3 + 1) - 3*y + log(y + 1) - log(y**2 - y + 1)/2 + sqrt(3)*atan(2*sqrt(3)*y/3 - sqrt(3)/3) + sqrt(3)*pi/6

>>>continuous_domain(p, y ,Interval(-10,5))
Interval.Lopen(-1, 5)
```
All though the function passed is same but calculating the integral first and then passing gives result.

smichr commented:
> the function passed is same but calculating the integral first and then passing gives result.

Hmm. I was thinking you would send a non-Integral example. Whoever wants to pass a univariate integrand in variable x with symbolic limits will have to think about what range of x will be of interest. In your example x goes from 0 to y and y from -10 to 5 so we are basically interested in x values in the range of -10 to 5...and I am guessing that if you pass the Integral with x (not y) as the variable of interest and Interval(10,5) as the range you will get the right answer.

jlherren committed:
Add assumption handlers for OneMatrix

jlherren committed:
Improve exceptions raised for matrices

- Raise more specific matrix exceptions whenever possible
- Raise correct exceptions on matrix inversion

jlherren committed:
Improve inversion of 2x2 block matrices

It previously tried to apply the formula even when A wasn't square.  It now
also considers invertibility of matrices to choose which formula to apply.

povinsahu1909 committed:
used continuous_domain

smichr committed:
int not Bool so Eq(bool, int) -> False

povinsahu1909 committed:
try except used with continuous_domain

mijo2 committed:
solvers: Removed the solver Linear, 3 equations, Order 1, Type 1

Removed the solver and ran tests to confirm that the new method
is computing the solution properly.

mijo2 committed:
solvers: Removing the solver Linear, 3 equations, Order 1, Type 2

Commit:
perm_groups.py: Add Coset class

Commit:
removed some changes

Commit:
changes example

asmeurer committed:
Merge pull request #19090 from sylee957/sphinx_deprecation

Change deprecated sphinx from sympylive

Commit:
added check for dir

povinsahu1909 commented:
> I am guessing that if you pass the Integral with x (not y) as the variable of interest and 
Interval(10,5) as the range you will get the right answer.

Sorry for the late response, as I am have problem with my system.
I used `try` and `except` with `continuous_domain` calculate domain using every variable present in an expression. The following is the code that I used
```python
for symbol in expr.atoms(Symbol):
     try:
          p = continuous_domain(expr, symbol, domain = Interval(x1, x2))
          cont_domain = Intersection(cont_domain , p)
     except:
          pass
```
Is this change valid? If it is then I will commit the changes.

smichr commented:
> Is this change valid? If it is then I will commit the changes.

I will check back later. Just happened to see this as I am leaving.

rj-calvin committed:
added test case for updated replace_with_arrays

smichr committed:
S.true != 1, too

jlherren committed:
Untangle ShapeError imports to be able to remove unused import

jlherren committed:
Use sympify in BlockDiagMatrix constructor

Smit-create committed:
test for labels

sylee957 committed:
Merge pull request #19093 from jlherren/blockmatrix

Various block matrix improvements

Commit:
perm_groups.py : Add relations in is_subgroup and __contains__ in SymmetricPermutationGroup

Commit:
add more tests and some minor improvements

smichr committed:
Merge pull request #19094 from smichr/linbis

add Line.bisectors

Commit:
added examples

Commit:
perm_groups.py : minor improvements

smichr committed:
Merge pull request #19054 from smichr/solvebool

be more specific when testing for T/F

smichr commented:
> Is this change valid? If it is then I will commit the changes.

I don't think so. The user should know what symbol they are interested in and the routine should be able to analyze the input to know what to test. Something like:
```
if isinstance(expr, Expr):
    check expr
elif isinstance(expr, Integral):
    figure out how passed symbol is related to free symbols of expr
etc..
else:
    raise NotImplementedError
```

Commit:
Merge pull request #18993 from mohitacecode/issue_18804

Solves key error in creating polycyclic group, Improves _sift, induced_pcgs

povinsahu1909 commented:
>The user should know what symbol they are interested in and the routine should be able to analyze the input to know what to test.

As this implementation is for 2D plot, in which if user plot multivariate function it will produce error.
So this 
```
p = continuous_domain(expr, symbol, domain = Interval(x1, x2))
``` 
will run for one time in most of the cases. But for the function which contain more than one symbol and are valid (i.e. function which contain Integral with symbolic range), this will run for number of symbol present in the passed function. And function will be plotted with the symbol passed.

>figure out how passed symbol is related to free symbols of expr

For this as we are using continuous_domain method which check continuity of the function based on the atomic value of the function rather than checking for the output function from the operation given, i.e. for example
```python
>>>continuous_domain(Integral(1/(x-1),(x,0,y)) , y, Interval(0,10))
Interval(0, 10)
>>>continuous_domain(Integral(1/(x-1),(x,0,y)) , x, Interval(0,10))
Union(Interval.Ropen(0, 1), Interval.Lopen(1, 10))
```
As the intersection of the above result gives interval `Union(Interval.Ropen(0, 1), Interval.Lopen(1, 10))`. But the desired is `Interval.Lopen(1, 10)`.
So if we somehow can get the domain after evaluating the expression i.e calculating the domain with w.r.t the final function, which in turn will result in the desired outcome without extra operations.

povinsahu1909 committed:
improved usage of continuous_domain method

jlherren committed:
Fix evaluation of inverse of inverse

Inverse(Inverse(A)).doit() was previously returning A ** 1.  This also moves
the responsibility of inverting a MatPow to the MatPow._eval_power method
instead of the Inverse class.

sylee957 committed:
Merge pull request #19079 from Smit-create/plot_label

Accept label as argument in plotting

oscarbenjamin committed:
Merge pull request #18877 from smitgajjar/py-upgrade2

Upgrading to Python3 (2) (pyupgrade to files from benchmarks to crypto)

sylee957 committed:
Use numpydoc format for plot

oscarbenjamin committed:
Merge pull request #19101 from sylee957/plot_doc

Improve plot docs

oscarbenjamin commented on sympy/plotting/tests/test_plot.py:
```diff
@@ -623,3 +635,30 @@ def test_issue_13516():
+        p.save(test_filename)
+        compare_images(cmp_filename , test_filename , 0.005)
+    finally:
+        TmpFileManager.cleanup()
```
Is it not possible to use with?

povinsahu1909 replied:
Yes, we can use with.
This test was added by taking #16674 as reference. But image comparison test is passing even if the file name is replaced with another file in the same directory.
So I planned to remove this test.

jlherren committed:
Forbid most powers of non-square matrices and code clean-up

This delegates special MatPow behavior to the respective classes instead of
handling them inside MatPow.doit().

ebraminio committed:
Remove two minor references to py2's print

sylee957 committed:
Merge pull request #19105 from ebraminio/minor-py2-rm

Remove two minor references to py2's print

jlherren committed:
Stabilize two flaky tests: test_pmint_erf/test_V12

They randomly failed due to itermonomials() yielding values in
non-deterministic order, because it uses set().

jlherren committed:
Fix combine_kronecker() manipulation for non-square matrices

(A x B) ** m is no longer transformed to A ** m x B ** m for non-square
matrices A, B, where x is the kronecker product.

tacaswell committed:
FIX: don't try to autoscale 3D axes

tacaswell committed:
MNT: Only autoscale if not manually set

jksuom committed:
Merge pull request #19087 from moorepants/mpl-less-than-3.2

Use matplotlib < 3.2 on Travis.

oscarbenjamin committed:
Merge pull request #19108 from jlherren/fix-unstable-tests

Stabilize two flaky tests: test_pmint_erf/test_V12

jlherren committed:
Raise nicer error when calling as_explicit() on symbolic-shaped matrix

mijo2 committed:
doc: Updated the file ode.rst

Removed the documentation details of the solvers that are to be removed
after adding the first general case solver

Commit:
added input checks for dir

mijo2 committed:
solvers: Updated the code

Some number of minor updates in the code were made.

Mohitbalwani26 committed:
refactored nth_algebraic examples and fixed a small bug in separable_reduced hint

Mohitbalwani26 committed:
fixed unused import

sbt4104 committed:
ZeroMatrix should not return False.

certik committed:
Doc: update the GSoC reports

asmeurer committed:
Merge pull request #19111 from certik/gsoc

Doc: update the GSoC reports

Mohitbalwani26 committed:
reverted back the fix for separable_reduced as it failed for doctest

sylee957 committed:
Merge pull request #19110 from sbt4104/master

ZeroMatrix should not return False.

mijo2 committed:
solvers: Added regression tests for issue #15474

mijo2 committed:
solvers: Moved test cases

Moved the test cases for the systems of ODEs in test_ode.py that are
now covered by the new general case solver.

mijo2 committed:
solvers: Updated the _get_func_order function

Updated the logic of _get_func_order function and minimised the
implmentation length

mijo2 committed:
solvers: Refactored linear_eq_to_matrix in systems.py

linear_eq_to_matrix function is used in systems.py and hence it is
refactored.

jlherren committed:
Completely disallow powers of non-square matrices

mijo2 committed:
solvers: Edited the import names

Mohitbalwani26 committed:
making one function call per line and fixed gen argument

jlherren committed:
Remove unnecessary method override

jlherren committed:
Fix bc_block_plus_ident() losing unrelated summands

smichr committed:
19113: public roots require symbol generator

The issue of solving cos(x)**5 - sin(x)**5 is circumvented
by a better solve strategy that rewrite sin/cos as tan via
TR2i. But the issue of roots being generated for expressions
with non-symbol generators is adjusted on two fronts:

1) public functions that do root-related work reject expressions
with non-symbol generators
2) univariate Poly which find roots will always return a canonical
RootOf with generator `x`.

sylee957 committed:
Merge pull request #19116 from jlherren/fix-bc_block_plus_ident

Fix bc_block_plus_ident() losing unrelated summands

RituRajSingh878 committed:
added new reduced_separable method in ode

Mohitbalwani26 committed:
reverted back the fix in ode.py as new PR was open for that

asmeurer committed:
Merge pull request #19006 from sylee957/sort_key_sympify

Deprecate sympify string fallback

Smit-create committed:
allow custom distributions

Smit-create committed:
fix failing tests

mijo2 committed:
solvers: Added linear_ode_to_matrix function

mijo2 committed:
solvers: Added regression test case for issue #15474

asmeurer commented:
Can we avoid adding image files for the tests (c.f. https://github.com/sympy/sympy/pull/18856#issuecomment-598872280)? Binary files can bloat the git history, so I would prefer not to check them in.

asmeurer commented on sympy/plotting/tests/test_plot.py:
```diff
@@ -623,3 +635,30 @@ def test_issue_13516():
+        cmp_filename = os.path.join(test_directory, "test_singularity2.png")
+        p = plot(sin(x)/x)
+        p.save(test_filename)
+        compare_images(cmp_filename , test_filename , 0.005)
```
I think we need to have a more general discussion about whether we want to do this sort of test in SymPy. For one thing, it requires checking in binary files to the repo, which can bloat the repo size. Another issue is that if matplotlib changes their plot format in some way, it would break the test. It would also happen if we change the way the plot is formatted in some way.  Is there some way we can test the plot without comparing binary files?

povinsahu1909 replied:
I added this type of test so that the extra line which get into plot can be compared with the desired plot. But currently, it is failing to produce test error.

> For one thing, it requires checking in binary files to the repo, which can bloat the repo size. 

This problem can be handled by a suggestion from @sylee957 who suggested to store the testing file to somewhere like sympy wiki. 
>Is there some way we can test the plot without comparing binary files?

I think we can do this by calculating the RMS value directly from the points which are used to plot the graph. This will remove our dependency on matplotlib for image comparison.

oscarbenjamin replied:
Can you query the line data from `p._backend.ax[0]` somehow?

povinsahu1909 replied:
As far as I understand the question yes we can find the line data from both the backend.

asmeurer reviewed (changes requested):
The issue of how to handle the tests for this needs to be addressed.

mijo2 committed:
solvers: Docstring fix

A bug was fixed in the dcocstring of the function linear_ode_to_matrix

mijo2 committed:
solvers: Added test cases for linear_ode_to_matrix

Smit-create committed:
tests and docs

RituRajSingh878 committed:
added test and modified code

RituRajSingh878 committed:
removed invalid syntax

jlherren committed:
Move code out of try..except block

Mohitbalwani26 committed:
added a check whether solution returned by dsolve without passing hint is correct

jlherren committed:
MatMul: Add and improve tests, properly call parent constructor

sylee957 committed:
PEP8 fixes

RituRajSingh878 committed:
updated test

sylee957 committed:
Merge pull request #19077 from mohitacecode/Coset_class

perm_groups.py: Add Coset class

smichr commented:
> But the desired is `Interval.Lopen(1, 10)`

is there a difference in the plot obtained?

sylee957 committed:
Merge pull request #19104 from jlherren/matpower

Clean up of MatPow code, forbid non-square matrix powers

sylee957 committed:
Merge pull request #19083 from jlherren/fix-cse-opts

Fix CSE to not construct invalid tree

mijo2 committed:
solvers: Updated the docstring of matrix_exp function

Updated the documentation of matrix_exp function in systems.py
according to the Sympy's documentation style guidelines

smichr committed:
Merge pull request #19117 from smichr/19113

19113: public roots require symbol generator

Commit:
add a Rule

Add this rule:  'Pochhammer[x,y]':'rf(x,y)'

Commit:
add Rules

add 'ArcTan[x,y]':'atan2(y,x)'

povinsahu1909 commented:
> is there a difference in the plot obtained?

No there will be no difference but due to this
>Union(Interval.Ropen(0, 1), Interval.Lopen(1, 10))

there is one extra function is created which is of no use.

Sc0rpi0n101 committed:
Merge pull request #19029 from smitgajjar/binary_op

parsing(c_parser): added full support for Binary Operators and any type of complicated Assignment in C parser

smichr committed:
enhance Mul.is_integer

RituRajSingh878 committed:
added checkodesol in a test

Mohitbalwani26 committed:
cleaning the code of newly added internal test functions

iammosespaulr committed:
Update cse_opts.py

Fixes #19136

mijo2 committed:
solvers: Updated docstring for function matrix_exp_jordan_form

mijo2 committed:
solvers: Small changes in some docstrings

oscarbenjamin committed:
Merge pull request #19122 from RituRajSingh878/reduced_ode

added new reduced_separable method in ode

oscarbenjamin committed:
Merge pull request #19063 from Mohitbalwani26/refactoring_tests

refactor(test_ode): Factorable examples

Commit:
Add test

Add test for mathematica paser.

Commit:
perm_groups and permutation : modify __mul__

Commit:
PEP8 fix

smichr committed:
Update cse_opts.py

smichr committed:
Apply suggestions from code review

smichr committed:
Merge pull request #19137 from iammosespaulr/patch-3

Update cse_opts.py

smitgajjar committed:
parsing(c_parser): added support for unary ops and paren expr
- Added support for unary operators:
    - unary plus(+), unary minus(-), pre/post increment(++), pre/post decrement(--) and logical NOT(!)
- Added support for parenthesized expression
- Added corresponding tests
    - Added tests for unary operators
    - Added tests for parenthesized expression
    - Added tests for Exceptions raised in parsing binary operators, variable declarations and unary operators
- Improved error messages of Exceptions raised in parsing binary operators, variable declarations and unary operators

Other notes:
- Dereferencing operator(*), Address operator(&) and bitwise NOT(~) are the unary operators that have not been implemented because of unavailability of such nodes or classes in SymPy
- If unary operators are the children of binary operator in AST, it is not possible to parse yet!

sylee957 committed:
Merge pull request #19139 from mohitacecode/modifymul

Define multiplication between Permutation and PermutationGroup

Commit:
delete __name__==__main__ ,instead put it outside

delete __name__==__main__

Commit:
delete __name__=="__main__"

delete __name__=="__main__"  ,instead, put it in test_mathematica.py

Commit:
Add test example

Add 'ArcTan[x,y]': 'atan2(y,x)' and  'Pochhammer[x,y]': 'rf(x,y)'.

jlherren committed:
PEP 8 fix

Co-Authored-By: Smit Gajjar <38134283+smitgajjar@users.noreply.github.com>

sylee957 committed:
Correct grammar

smichr committed:
Merge pull request #19060 from Psycho-Pirate/19059

Fixed error with Pyglet Plotting

smichr committed:
Merge pull request #19142 from jlherren/as-explicit

Raise nicer error when calling as_explicit() on symbolic-shaped matrix

sylee957 committed:
Merge pull request #19135 from wuyudi/master

add 2 rules in mathematica parser

akirakyle committed:
add _eval_adjoint and _eval_transpose to ExpBase

mijo2 committed:
solvers: Added NonLinearError in solveset

mijo2 committed:
solvers: Added ODENonLinearError in systems.py

mijo2 committed:
solvers: Updated a doctest in linsolve

A doctest in function linsolve was giving incorrect results. It was
updated.

mijo2 committed:
solvers: Added canonical equations function

iammosespaulr committed:
add support for binom in latex

jlherren committed:
Improve printing of matrix slices

iammosespaulr committed:
fixes

sylee957 committed:
Merge pull request #19151 from akirakyle/master

add _eval_adjoint and _eval_transpose to ExpBase

smichr committed:
enhance Pow.is_integer

smichr committed:
add assumption helpers; relocate

smichr committed:
flake8 mods to Mul

smichr committed:
add note about storage of assumptions

mijo2 committed:
solvers: Updated the name of NonLinearError to NonlinearError

mijo2 committed:
solvers: Updated canonical_equations function

Made canonical_equations private.

mijo2 committed:
solvers: Updated the solver

The solver can now collect exp terms together.

iammosespaulr committed:
downgrade to 4.7.1

mijo2 committed:
solvers: Added regression test case for issue #19150

Upabjojr committed:
Merge pull request #19091 from cjayross/tensor_contraction

Fixed tensor contraction problem

jlherren committed:
Improve 2x2 block matrix inversion by implementing all formulas

sylee957 committed:
Merge pull request #19147 from iammosespaulr/origin/latex

add support for binom in latex

smichr committed:
Merge pull request #19155 from smichr/assumptions

adding to and moving assumptions

smitgajjar committed:
ran pyupgrade across files ranging from deprecated to ntheory module

smitgajjar committed:
added coding utf-8 to resolve travis fail

smitgajjar committed:
added coding utf-8

smitgajjar committed:
trigger travis

smichr committed:
19157 recognize when quintics fails

smichr committed:
replace nonsym collectables with symbols

mijo2 committed:
solvers: Updated some of the test cases in test_ode.py

Updated the test cases and added a required comment for future
reference

mijo2 committed:
doc: Updated the path for the general case solver

mijo2 committed:
solvers: Updated the incomplete documentation in function linear_ode_to_matrix

mijo2 committed:
solvers: Added some minor changes requested.

mijo2 committed:
solvers: Added a workaround for the collect function

Collect function doesn't work correctly for some cases, (more information
in issue #19149). A workaround is added in our general case solver.

mijo2 committed:
doc: Added documentation details for some functions in systems.py

Added documentation details for new functions like linear_ode_to_matrix,
matrix_exp_jordan_form, matrix_exp.

mijo2 committed:
solvers: Updated doc files for successful build.

smichr committed:
Merge pull request #19165 from smichr/19149

collecting nonsymbols; solving quintics

oscarbenjamin committed:
Merge pull request #19160 from smitgajjar/pyupgrade3

Upgrading to python3 (3)

sylee957 committed:
Fix numpy printing of unevaluated pow

mijo2 committed:
Merge branch 'master' into 15549_matrix_exp

mijo2 committed:
solvers: Removing the workaround in systems.py

oscarbenjamin committed:
Merge pull request #19170 from sylee957/fix_numpy_pow

Fix numpy printing of unevaluated pow

oscarbenjamin committed:
Merge pull request #18720 from mijo2/15549_matrix_exp

Using matrix exp for solving linear differential equations

Mohitbalwani26 committed:
refactoring riccati_special_minus2 test

oscarbenjamin committed:
Merge pull request #19176 from Mohitbalwani26/refactoring_tests

refactor(test_ode): Riccati_special_minus2

smichr committed:
Merge pull request #8348 from cbm755/not_all_basic

Basic: docstring: not all objects in SymPy are Basic

MartinThoma committed:
DOC: Use README.rst for pypi.org page

iammosespaulr committed:
Okay!

Mohitbalwani26 committed:
refactor(test_ode.py): regression tests

Mohitbalwani26 committed:
marked examples as xfail for the hints which they were failing

MartinThoma committed:
Convert README to markdown with pandoc

czgdp1807 committed:
Merge pull request #19123 from Smit-create/custom_dists

Allow custom probability distributions

czgdp1807 committed:
removed user guide

Mohitbalwani26 committed:
added _test_dummy_sol function to handle solutions with dummy variable and shifted most of try block outside as it had bugs

Mohitbalwani26 committed:
import fix

sylee957 committed:
Change plot_parametric docs to numpydoc format

oscarbenjamin committed:
Merge pull request #19179 from MartinThoma/pypi

DOC: Use README.rst for pypi.org page

MartinThoma committed:
STY: Use Python 3 syntax in vector module

sylee957 committed:
Improve the notes

sylee957 committed:
Add plot example for notes

iammosespaulr committed:
upgrade Mul.is_integer and fix tests

Co-authored-by: Chris Smith <smichr@gmail.com>

fix failing

czgdp1807 committed:
removed unwanted pics

oscarbenjamin committed:
Merge pull request #19194 from MartinThoma/vector-py3

Use Python 3 syntax in vector module

smichr committed:
Merge pull request #19182 from iammosespaulr/assump

fix Mul.is_integer

jlherren committed:
Always print matrix slices in a way to not be confused with matrix elements

sylee957 committed:
Do not attempt to factor if the expression has float

sylee957 committed:
Update sympy/plotting/plot.py

Co-Authored-By: Christopher Smith <smichr@gmail.com>

oscarbenjamin committed:
Merge pull request #18985 from sympy/revert-18434-17119

Revert "Handling Float Integrals"

sylee957 committed:
Merge pull request #19201 from jlherren/slice-printing

Improve matrix slice printing

sylee957 committed:
Fix flake8

sylee957 committed:
Improve docs for xlim and ylim

sylee957 committed:
Improve example description

sylee957 committed:
Merge branch 'improve_plot_docs' of https://github.com/sylee957/sympy into improve_plot_docs

sylee957 committed:
Fix typo

sylee957 committed:
Merge pull request #19088 from tacaswell/fix_autoscale

FIX: ensure autoscale_view gets called

sylee957 committed:
Remove division by 2 for generalized cases

sylee957 committed:
Use Gershgorin circle theorem for testing positive definite

sylee957 committed:
Fix wrong absolute value computation

sylee957 committed:
Fix typo

sylee957 committed:
Fix typo

sylee957 committed:
Return false for nonsquare cases

sylee957 committed:
Return false when diagonals are strictly negative

sylee957 committed:
Fix wrong stats test cases

aktech committed:
Update all LICENSE Copyright year

sylee957 committed:
Fix invalid latex control

sylee957 committed:
Change index notation

Commit:
Merge pull request #19213 from sympy/license-2020

Update all LICENSE Copyright year

sylee957 committed:
Use more cheaper test for solvers

sylee957 committed:
Add regression test for #18955

Mohitbalwani26 committed:
simplified the code for _test_paticular_example_

Mohitbalwani26 committed:
_test_an_example function calls test_particular_example for the code output

jlherren committed:
Remove unused import

eric-wieser committed:
Speed up Symbol.sort_key

There's no need to go through `StrPrinter` here.

eric-wieser committed:
Speed up Dummy.sort_key

There's no need to go through `StrPrinter` here.

asmeurer committed:
Merge pull request #19190 from czgdp1807/user_guide

User guide removed

smichr committed:
Merge pull request #19216 from eric-wieser/patch-5

Speed up Symbol.sort_key

aktech committed:
Make copyright in docs dynamic

Mohitbalwani26 committed:
refactored liouville test examples

Sc0rpi0n101 committed:
Merge pull request #19140 from smitgajjar/unary_op

parsing(c_parser): added support for unary operators and outermost parenthesized expression

asmeurer committed:
Put the doctr environment variable only in the build that needs it

Aside from being more secure, this fixes the environment variable listing on
the Travis site from showing DOCTR_DEPLOY_ENCRYPTION_KEY=[secure] as the first
thing for every build.

asmeurer committed:
Order the environment variables better in .travis.yml

The environment variables are shown on the build listing on the Travis site in
the order they are in the file, so it's best to put them in a consistent order
with the most important variables first.

sylee957 committed:
Merge pull request #19226 from asmeurer/doctr-env

Put the doctr environment variable only in the build that needs it

sylee957 committed:
Merge pull request #19219 from aktech/dynamic-copyright

Make copyright in docs dynamic

sylee957 committed:
Inherit MatrixExpr for immutable sparse matrix

sylee957 committed:
Add tests

sylee957 committed:
Fix test

oscarbenjamin committed:
Merge pull request #19205 from sylee957/improve_positive_definite

Use gaussian elimination for positive-definite check

oscarbenjamin committed:
Merge pull request #19228 from sylee957/fix_immutable_sparse

Inherit MatrixExpr for ImmutableSparseMatrix

sylee957 committed:
Fix creation of matrix with list of ndarray

sylee957 committed:
Revert redundant change

sylee957 committed:
Fix test without numpy

oscarbenjamin committed:
Merge pull request #19229 from sylee957/fix_ndarray_matrix

Fix creation of matrix with list of ndarray

oscarbenjamin committed:
Remove expand_mul call in cholesky and LDLdecompisition

Both cholesky and LDLdecompisition used expand_mul to simplify some
calculations but this can lead to significant slowdowns in some cases.
This commit removes the calls to expand_mul from each of these
functions. A few tests needed to be changed because the output is no
longer simplified to the same extent.

See https://github.com/sympy/sympy/issues/19200

sylee957 committed:
Merge pull request #19215 from jlherren/improve-2x2-block-inversion

Improve 2x2 block matrix inversion with off-diagonal invertible blocks

eric-wieser committed:
Respect `inv_trig_style` in inverse hyperbolic trig functions

Upstreamed from the galgebra printer, which seemed to want these.

oscarbenjamin committed:
Merge pull request #19184 from Mohitbalwani26/refatoring_tests

refactor(test_ode.py): regression tests

czgdp1807 commented:
Any news on this @povinsahu1909 ?

povinsahu1909 commented:
I'll commit the changes that I have have made and I also thought of removing the image comparison test which I added as it was not effective.
And sorry for the late reply as I am having network issue.

sylee957 committed:
Merge pull request #19235 from eric-wieser/include-hyperbolic-in-inv_trig

Respect `inv_trig_style` in inverse hyperbolic trig functions

sylee957 committed:
Display examples for plot_implicit

sylee957 committed:
Fix doctest

oscarbenjamin committed:
Merge pull request #19239 from sylee957/plot_implicit_examples

Display examples for plot_implicit

oscarbenjamin committed:
Merge pull request #19233 from oscarbenjamin/pr_expand_mul_revert

Remove expand_mul call in cholesky and LDLdecompisition

oscarbenjamin committed:
Bump version to 1.7.dev

oscarbenjamin committed:
Merge pull request #19244 from oscarbenjamin/pr_bump_1.7.dev

Bump version to 1.7.dev

smichr committed:
Merge pull request #19192 from sylee957/improve_plot_docs

Improve plot_parametric docs

povinsahu1909 committed:
updated 2d plot functon

povinsahu1909 commented:
Anyone please help me, I completely messed up this pr :(

asmeurer commented:
@povinsahu1909 it looks like this might be a rebase gone bad. 

First, if you have any uncommitted changes, use `git stash` to get rid of them. You can put them back after this is all done with `git stash pop`. 

To fix it, first you'll want to make a backup branch, in case you do any of the steps below wrong

```
git checkout -b discontuinity-backup
git checkout discontuinity
```

(you spelled discontinuity wrong in your branch name btw)

Then make sure your `master` branch is the correct one. 

Run. 

```
git remote show origin
```

If this shows your fork instead of sympy/sympy, run

```
git remote set-url origin git@github.com:sympy/sympy.git
```

Then run

```
git checkout master
git fetch origin
git branch --set-upstream-to=origin/master master
git reset --hard master
```

This will make sure that master is up-to-date with the SymPy master, and tracks it. In the future, you can update master by running

```
git checkout master
git pull
```

Then run

```
git checkout discontuinity
git rebase -i master
```

This will open a text editor with all the commits. You should remove every commit from the list except for yours (delete the lines), and save the file. Git will then fix the branch. If there are any errors or issues, it will tell you.

After this is done, assuming it is successful, you can push it up with `git push -f <remote> discontuinity` where `<remote>` is whatever you called your fork remote (usually either `github` or `povinsahu1909`). If you don't have a fork remote add it with

```
git remote add povinsahu1909 git@github.com:povinsahu1909/sympy.git
```

The `-f` flag is required to push it up since you rebased. 

In the future, I would avoid rebasing, as it can lead to issues like this. Instead, if there are merge conflicts or if you need changes from master, use `git merge`. `git merge` will never lead to issues like this and doesn't require using `-f` to push.

povinsahu1909 commented:
I have done all the above steps but I am unable to reopen the pr. And in the text editor as previously, it is only showing my commits, not the commits that are reflected here.
